Add BIP152 (Compact Blocks) structures

0xB10C opened the pull request:
> Adds the basic structures for BIP152 and a method to create a compact block from a full block.

This is a rebase of #249 by stevenroose (see https://github.com/rust-bitcoin/rust-bitcoin/pull/249#issuecomment-1170562141) with a milestone for 29.0. I've added deserialization and serialization tests for the network messages and fixed an off-by-one bug in the deserialization of the differentially encoded varints in the `getblocktxn` message (see https://github.com/rust-bitcoin/rust-bitcoin/pull/249#discussion_r914989521).  

Closes #249.

0xB10C commented:
@tcharding you had two outstanding comments in #249: https://github.com/rust-bitcoin/rust-bitcoin/pull/249#discussion_r824219887 and https://github.com/rust-bitcoin/rust-bitcoin/pull/249#discussion_r824220598. If they are still relevant, feel free to re-add them here and I'll address them.

apoelstra reviewed (approved):
ACK ea50200939b5296c350cf04615fa932cb66c6fb9

tcharding commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,476 @@
+    }
+}
+
+impl hex::FromHex for ShortId {
+    fn from_byte_iter<I>(iter: I) -> Result<Self, hex::Error>
+    where
+        I: Iterator<Item = Result<u8, hex::Error>> + ExactSizeIterator + DoubleEndedIterator,
+    {
+        Ok(ShortId(hex::FromHex::from_byte_iter(iter)?))
+    }
+}
+
+impl fmt::LowerHex for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        hex::format_hex(&self.0[..], f)
+    }
+}
+
+impl fmt::Display for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl fmt::Debug for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
```
These can be replaced with `impl_bytes_newtype!(ShortId, 6);`

tcharding replied:
Note, doing so would add three more implementations as well, `FromStr`, `Serialize`, and `Deserialize`. Not sure right now if that's ok so flagging it.

apoelstra replied:
I think the extra impls are probably good ideas.

tcharding reviewed (commented):
I can't remember right now why I suggested adding `AsRef`, I'll do a more full review and see if it comes out of my brain.

0xB10C commented:
Added `impl_bytes_newtype!()` (and `impl_array_newtype!()`) for `ShortId` as discussed in https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r917333117. Invalidated ACK from @apoelstra.

tcharding commented:
The license blurb at the top of the new file needs changing now that https://github.com/rust-bitcoin/rust-bitcoin/pull/1076 merged. Sorry about that :)

tcharding reviewed (commented):
Two more small things, then looks good to go! Thanks for doing this man.

tcharding commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,451 @@
+
+/// Short transaction IDs are used to represent a transaction without sending a full 256-bit hash.
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Default, PartialOrd, Ord)]
+pub struct ShortId(pub [u8; 6]);
```
I'm basing this suggestion off the other three uses in the lib of `impl_array_newtype`.

0xB10C commented:
Addressed https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r918444920 and https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r918446368. Thank you for reviewing and helping to move this forward.

apoelstra reviewed (dismissed):
ACK 9acd3de75ac5a65717fa9a0ae95b898a5df09bec

tcharding reviewed (dismissed):
ACK 9acd3de75ac5a65717fa9a0ae95b898a5df09bec

apoelstra commented:
This is by far our oldest PR (if you consider it to be the same as #249) ... I feel a bit weird hitting Merge on it :)

0xB10C commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+                let mut last_index: u64 = 0;
+                for _ in 0..nb_indexes {
+                    let differential: VarInt = Decodable::consensus_decode(&mut d)?;
+                    last_index = match last_index.checked_add(differential.0) {
+                        Some(r) => r,
+                        None => return Err(encode::Error::ParseFailed("block index overflow")),
+                    };
+                    indexes.push(last_index);
+                    last_index += 1;
```
I think we should use `checked_add` in L289 too. L289 panics `'attempt to add with overflow', src/util/bip152.rs:289:21` with an input of `vec![1, 255, 254, 255, 255, 255, 255, 255, 255, 255]` (differentially encoded `VarInt(u64::MAX)` with length).

0xB10C replied:
Will fix this up later and add a test case to `test_getblocktx_differential_encoding_de_and_serialization`. Sorry for invalidating your ACKs again..

RCasatta commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+    fn consensus_decode<D: io::Read + ?Sized>(mut d: &mut D) -> Result<PrefilledTransaction, encode::Error> {
+        let idx = VarInt::consensus_decode(&mut d)?.0;
+        if idx > u16::max_value() as u64 {
+            //TODO use TryInto instead when MSRV exceeds 1.34
```
implement TODO since we are now on 1.41.1?

tcharding replied:
Yep, this works
```rust
        let idx = VarInt::consensus_decode(&mut d)?.0;
        let idx = u16::try_from(idx).map_err(|_| encode::Error::ParseFailed("BIP152 prefilled tx index out of bounds"))?;
        let tx = Transaction::consensus_decode(&mut d)?;

        Ok(PrefilledTransaction { idx, tx })
```

RCasatta commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+
+        let siphash_keys = ShortId::calculate_siphash_keys(&block.header, nonce);
+
+        let mut prefilled = vec![];
+        let mut short_ids = vec![];
```
I think we can initialize the capacity of this vectors by using `prefill.len()` and `block.txdata.len()`

junderw replied:
Specifically:

```rust
let mut prefilled = Vec::with_capacity(prefill.len() + 1); // +1 for coinbase tx
let mut short_ids = Vec::with_capacity(block.txdata.len() - prefill.len() - 1);
```

0xB10C replied:
I'm not sure why you subtract one in the `short_ids` capacity. This would result in a negative capacity when all transactions are pre-filled (catched by testcase 2 in `test_header_and_short_ids_from_block`). 

I think `short_ids` should be: 
```
let mut short_ids = Vec::with_capacity(block.txdata.len() - prefill.len());
```

junderw replied:
hmmm, so it's possible for the coinbase tx to be inside the prefill vec?

I was judging from the logic where you were `prefilled.push()` when index == 0, that coinbase tx would never be passed by the prefill vec and would always be fetched from block.txdata.

Is this behavior documented somewhere?

`prefill can contain coinbase tx or not, but either way, the coinbase tx will be treated as if it is prefilled` etc.

0xB10C replied:
I think the coinbase transaction is always pre-filled with BIP152. The receiving peer has no way of knowing the coinbase transaction before it knows the block. 

From BIP 125 in the purpose column of `prefilledtxn`:

> Used to provide the coinbase transaction and a select few which we expect a peer may be missing

https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#headerandshortids

junderw commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+use crate::util::endian;
+
+/// A BIP-152 error
+#[derive(Clone, PartialEq, Eq, Debug, Copy, PartialOrd,Ord, Hash)]
```
nit: missing space

junderw commented:
LGTM, besides 1 nit and a `+1` for all of the unresolved comments.

0xB10C commented on src/util/bip152.rs:
```diff
@@ -0,0 +1,441 @@
+        let mut last_idx = 0;
+        for idx in &self.indexes {
+            len += VarInt(*idx - last_idx).consensus_encode(&mut s)?;
+            last_idx = *idx + 1;
```
I'm not sure how to handle a error during serialization. I've added a overflow check here and return an `io:ErrorKind::InvalidData` but this still panics: `panicked at 'in-memory writers don't error: Custom { kind: InvalidData, error: "block index overflow" }', src/consensus/encode.rs:145:51`...

0xB10C commented:
- addressed https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r919849367: no longer panic in deserialization of `BlockTransactionRequest`. Also added two testcases for this.
- as mentioned in https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r923887170, documented that serialization can panic if the indexes vector includes `u64::MAX`. Docs on both the `BlockTransactionRequest::indexes` field and the `consensus_encode()` impl. Lmk if that's OK.
- implemented the TODO mentioned in https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r919864644
- addressed https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r919874845 based on @junderw's suggestion but used https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r924248528 instead
- ran `rustfmt on src/util/bip152.rs` to address https://github.com/rust-bitcoin/rust-bitcoin/pull/1088#discussion_r922989132 and to ease https://github.com/rust-bitcoin/rust-bitcoin/pull/1040
- rebased on lastest master which meant:
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1065: replaced the `BlockHash::default()` occurrences
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1093: use `Sequence()`

apoelstra reviewed (dismissed):
ACK 8d06fd1a790edcf8ab867496ba500a0095840f3e

tcharding reviewed (dismissed):
ACK 8d06fd1a790edcf8ab867496ba500a0095840f3e

tcharding commented:
Super thorough, thanks @0xB10C!

apoelstra commented:
This needs a rebase to `use crate::hashes` -- currently it does not build after merging.

tcharding commented:
Ha that's funny, I couldn't for the life of me work out why it built locally without `crate` in the import line of `hashes`. I just thought I must have been being brain dead.

0xB10C commented:
Rebased. 

Added `crate` to `use hashes` and added the `internal_macros` imports required since https://github.com/rust-bitcoin/rust-bitcoin/pull/1041/commits/21a1cc791cebd4bd1e83c9b74b50e947668c7d31.

tcharding reviewed (approved):
ACK ee29910911e786e40b1cb25c72ed90eaad48c512

apoelstra reviewed (approved):
ACK ee29910911e786e40b1cb25c72ed90eaad48c512

apoelstra commented:
I think 3.5 years is long enough for this PR to get in :)
